dir.c

```diff
@@ -1397,6 +1397,17 @@ enum pattern_match_result path_matches_pattern_list(
 	strbuf_addch(&parent_pathname, '/');
 	strbuf_add(&parent_pathname, pathname, pathlen);
 
+	/*
+	 * Directory entries are matched if and only if a file
+	 * contained immediately within them is matched. For the
+	 * case of a directory entry, modify the path to create
+	 * a fake filename within this directory, allowing us to
+	 * use the file-base matching logic in an equivalent way.
+	 */
+	if (parent_pathname.len > 0 &&
+	    parent_pathname.buf[parent_pathname.len - 1] == '/')
+		strbuf_add(&parent_pathname, "-", 1);
```

Comment:
For some reason having a filename as `.` seems stranger to me than `-`. Perhaps `_` would seem less odd?

Reply:
I talked with @dscho offline here and here is the conclusion:

* The recommendation here is clever: `path/to/directory/.` is semantically identical to `path/to/directory`, so `.` is a natural option here.
* This relies on the cone-mode path-checking logic to be ignorant of this meaning of the trailing `/.`. While it _is_ dumb enough to ignore that, we would be adding an expectation that it remains dumb and otherwise doesn't check that the input path is a valid name for a file.

The conclusion is that this is a clever idea, but perhaps _too clever_.

Reply:
I'm not sure there is a correct answer here.  But it makes me want to say that we want to treat the ce's as a {type,path} tuple with an accessor and all that and get away from the pattern of just iterating over a dumb list of strings (that all callers know is a dumb list of strings).

Is there existing code to detect file -vs- symlink collisions?  or file -vs- submodule collisions?  How are they handled?
I mean we're adding the concept of (sparse) directory nodes into the index; it would be nice if they collided the same way.  (I'm not trying to cause problems, just asking dumb questions.)